python/nx-cugraph/scripts/update_readme.py

```diff
@@ -0,0 +1,203 @@
+    ),
+}
+
+
+def main(readme_file, objects_filename):
+    """``readme_file`` must be readable and writable, so use mode ``"a+"``"""
+    # Use the `objects.inv` file to determine URLs. For details about this file, see:
+    # https://sphobjinv.readthedocs.io/en/stable/syntax.html
+    # We might be better off using a library like that, but roll our own for now.
+    with Path(objects_filename).open("rb") as objects_file:
+        line = objects_file.readline()
+        if line != b"# Sphinx inventory version 2\n":
+            raise RuntimeError(f"Bad line in objects.inv:\n\n{line}")
+        line = objects_file.readline()
+        if line != b"# Project: NetworkX\n":
+            raise RuntimeError(f"Bad line in objects.inv:\n\n{line}")
+        line = objects_file.readline()
+        if not line.startswith(b"# Version: "):
+            raise RuntimeError(f"Bad line in objects.inv:\n\n{line}")
+        line = objects_file.readline()
+        if line != b"# The remainder of this file is compressed using zlib.\n":
+            raise RuntimeError(f"Bad line in objects.inv:\n\n{line}")
+        zlib_data = objects_file.read()
+    objects_text = zlib.decompress(zlib_data).decode().strip()
+    objects_list = [parse_docobject(line) for line in objects_text.split("\n")]
+    doc_urls = {
+        obj.name: "https://networkx.org/documentation/stable/" + obj.uri
+        for obj in objects_list
+    }
+    if len(objects_list) != len(doc_urls):
+        raise RuntimeError("Oops; duplicate names found in objects.inv")
+
+    def get_payload(info, **kwargs):
+        path = "networkx." + info.networkx_path
+        subpath, name = path.rsplit(".", 1)
+        # Many objects are referred to in modules above where they are defined.
+        while subpath:
+            path = f"{subpath}.{name}"
+            if path in doc_urls:
+                return f'<a href="{doc_urls[path]}">{name}</a>'
+            subpath = subpath.rsplit(".", 1)[0]
+        warn(f"Unable to find URL for {name!r}: {path}", stacklevel=0)
+        return name
+
+    def get_payload_internal(keys):
+        path = "networkx." + ".".join(keys)
+        name = keys[-1]
+        if path in doc_urls:
+            return f'<a href="{doc_urls[path]}">{name}</a>'
+        path2 = "reference/" + "/".join(keys)
+        if path2 in doc_urls:
+            return f'<a href="{doc_urls[path2]}">{name}</a>'
+        if path in MANUAL_OBJECT_URLS:
+            return f'<a href="{MANUAL_OBJECT_URLS[path]}">{name}</a>'
+        warn(f"Unable to find URL for {name!r}: {path}", stacklevel=0)
+        return name
+
+    readme_file.seek(0)
+    text = readme_file.read()
+    tree = create_tree(get_payload=get_payload)
+    # Algorithms
+    match = re.search(
+        r"### .Algorithms(?P<preamble>.*?)<pre>\n(?P<body>.*?)\n</pre>",
+        text,
+        re.DOTALL,
+    )
+    if not match:
+        raise RuntimeError("Algorithms section not found!")
```

Comment:
Personal preference, feel free to ignore: I like being less sensitive to the current README contents and just dumping all the generated content in a large predefined section vs. surgically inserting smaller sections. This feels more fragile for no obvious gain, but I don't feel strongly either way.

Reply:
The current method allows us to add more notes between the section header and the table, which I think we'll probably want to do. The current surgical regexes doesn't seem particularly fragile to me and I think will be easy to evolve.